src/components/GradeCalculator.tsx

```diff
@@ -0,0 +1,78 @@
+import { Course, Grade } from 'api';
+
+function normalize(grade?: Grade) {
+	if (typeof grade === 'number') {
+		return grade / 100;
+	}
+	if (typeof grade === 'undefined') {
+		return 0;
+	}
+	if (grade.denominator === 0) {
+		return 0;
+	}
+	return grade.numerator / grade.denominator;
+}
+
+function currentGradeCalculator(course: Course) {
+	//earnedGrade / sum of weights with grades
+	return (
+		earnedGradeCalculator(course) /
+		course.items
+			.filter((item) => item.grade != undefined)
+			.reduce((acc, item) => acc + normalize(item.weight), 0)
+	);
+}
+
+function earnedGradeCalculator(course: Course) {
+	//Grades * Weights
+	const normalizedGrade = course.items.reduce((acc, item) => {
+		return acc + normalize(item.weight) * normalize(item.grade);
+	}, 0);
+
+	const totalWeight = course.items.reduce((acc, item) => {
+		return acc + normalize(item.weight);
+	}, 0);
+	return (normalizedGrade / totalWeight) * 100;
+}
+
+function lostGradeCalculator(course: Course) {
+	//Sum of (weight - (grade * weight))
+	return (
+		course.items
+			.filter((item) => item.grade != undefined)
+			.reduce((acc, item) => {
+				return (
+					acc +
+					(normalize(item.weight) -
+						normalize(item.grade) * normalize(item.weight))
+				);
+			}, 0) * 100
+	);
+}
+
+function averageGradeNeededCalculator(course: Course) {
+	//Gets sum of weights of course items with grades not defined yet
+	const gradeRemaining = course.items
+		.filter((item) => item.grade === undefined)
+		.reduce((acc, item) => {
+			return acc + item.weight;
+		}, 0);
+	if (course.desiredGrade === undefined) {
+		return 0;
+	}
+	//Calculates average grades needed on rest of course items to get desired grade
+	const gradeNeeded =
+		(earnedGradeCalculator(course) - course.desiredGrade) /
+		(-1 * gradeRemaining);
+
+	if (gradeNeeded < 0) {
+		return 0;
+	} else return gradeNeeded * 100;
```

Comment:
We should omit the else statement and put the return on the next line

```suggestion
	if (gradeNeeded < 0) {
		return 0;
	}
	return gradeNeeded * 100;
```